httpx/_client.py

```diff
@@ -1628,6 +1654,8 @@ async def aclose(self) -> None:
             if proxy is not None:
                 await proxy.aclose()
 
+        self._is_closed = True
```

Comment:
We should put `self._is_closed = True` as the very first line, rather than the very last line, since that'd handle potential race conditions more cleanly.

Reply:
@tomchristie It a bit confuses me, since there should be another ways to avoid race conditions, e.g. locks.

Reply:
I saw several same places where race condition _is possible_ 

I can try to write a test, which strictly cause a race condition here